plugin/ready/setup.go

```diff
@@ -0,0 +1,51 @@
+package ready
+
+import (
+	"net"
+
+	"github.com/coredns/coredns/plugin"
+
+	"github.com/mholt/caddy"
+)
+
+func init() {
+	caddy.RegisterPlugin("ready", caddy.Plugin{
+		ServerType: "dns",
+		Action:     setup,
+	})
+}
+
+func setup(c *caddy.Controller) error {
+	addr, err := parse(c)
+	if err != nil {
+		return plugin.Error("ready", err)
+	}
+
+	r := new(addr)
+
+	c.OnStartup(r.onStartup)
+	c.OnStartup(r.wait)
+	c.OnRestart(r.onRestart)
+	c.OnFinalShutdown(r.onFinalShutdown)
+
```

Comment:
Thanks to the new "OnRestartFailed" event that was introduced few months ago in Caddy.
We can handle the failed restart (when the Corefile is invalid).

i guess we could have:
```
        c.OnStartup(r.onStartup)
	c.OnStartup(r.wait)
	c.OnRestart(r.onRestart)
         c.OnRestartFailed(r.onStartup)
	c.OnRestartFailed(r.wait)
	c.OnFinalShutdown(r.onFinalShutdown)
```